File: ai-agent-server/src/main/kotlin/com/asakii/server/mcp/UserInteractionMcpServer.kt
package com.asakii.server.mcp

import com.asakii.ai.agent.sdk.AiAgentProvider
import com.asakii.ai.agent.sdk.McpSystemPromptContext
import com.asakii.claude.agent.sdk.mcp.McpServerBase
import com.asakii.claude.agent.sdk.mcp.ToolResult
import com.asakii.claude.agent.sdk.mcp.annotations.McpServerConfig
import com.asakii.rpc.proto.AskUserQuestionRequest
import com.asakii.rpc.proto.QuestionItem as ProtoQuestionItem
import com.asakii.rpc.proto.QuestionOption as ProtoQuestionOption
import com.asakii.server.mcp.schema.SchemaValidator
import com.asakii.server.mcp.schema.ValidationResult
import com.asakii.server.rpc.ClientCallerRegistry
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import kotlinx.serialization.json.jsonPrimitive
import com.asakii.logging.*

@Serializable
data class UserAnswerItem(
    val question: String,
    val header: String,
    val answer: String
)

@Serializable
enum class PermissionBehavior {
    @kotlinx.serialization.SerialName("allow")
    ALLOW,
    @kotlinx.serialization.SerialName("deny")
    DENY,
    @kotlinx.serialization.SerialName("ask")
    ASK
}

@Serializable
enum class PermissionMode {
    @kotlinx.serialization.SerialName("default")
    DEFAULT,
    @kotlinx.serialization.SerialName("acceptEdits")
    ACCEPT_EDITS,
    @kotlinx.serialization.SerialName("plan")
    PLAN,
    @kotlinx.serialization.SerialName("bypassPermissions")
    BYPASS_PERMISSIONS
}

@Serializable
enum class PermissionUpdateDestination {
    @kotlinx.serialization.SerialName("userSettings")
    USER_SETTINGS,
    @kotlinx.serialization.SerialName("projectSettings")
    PROJECT_SETTINGS,
    @kotlinx.serialization.SerialName("localSettings")
    LOCAL_SETTINGS,
    @kotlinx.serialization.SerialName("session")
    SESSION
}

@Serializable
enum class PermissionUpdateType {
    @kotlinx.serialization.SerialName("addRules")
    ADD_RULES,
    @kotlinx.serialization.SerialName("replaceRules")
    REPLACE_RULES,
    @kotlinx.serialization.SerialName("removeRules")
    REMOVE_RULES,
    @kotlinx.serialization.SerialName("setMode")
    SET_MODE,
    @kotlinx.serialization.SerialName("addDirectories")
    ADD_DIRECTORIES,
    @kotlinx.serialization.SerialName("removeDirectories")
    REMOVE_DIRECTORIES
}

@Serializable
data class PermissionRuleValue(
    val toolName: String,
    val ruleContent: String? = null
)

@Serializable
data class PermissionUpdate(
    val type: PermissionUpdateType,
    val rules: List<PermissionRuleValue>? = null,
    val behavior: PermissionBehavior? = null,
    val mode: PermissionMode? = null,
    val directories: List<String>? = null,
    val destination: PermissionUpdateDestination? = null
)

@Serializable
data class PermissionResponse(
    val approved: Boolean,
    val permissionUpdates: List<PermissionUpdate>? = null,
    val denyReason: String? = null
)

@Serializable
data class AskUserQuestionParams(
    val questions: List<QuestionItem>
)

@Serializable
data class QuestionItem(
    val question: String,
    val header: String? = null,
    val options: List<OptionItem>? = null,
    val multiSelect: Boolean = false
)

@Serializable
data class OptionItem(
    val label: String,
    val description: String = ""
)

@Serializable
data class ClaudeQuestionItem(
    val question: String,
    val header: String? = null,
    val options: List<ClaudeOptionItem>? = null,
    val multiSelect: Boolean = false
)

@Serializable
data class ClaudeOptionItem(
    val label: String,
    val description: String = ""
)

private val mcpLogger = getLogger("UserInteractionMcpServer")

@McpServerConfig(
    name = "user-interaction",
    version = "1.0.0",
    description = "用户交互工具服务器，提供向用户提问等功能"
)
class UserInteractionMcpServer : McpServerBase() {
    private var timeoutMs: Long? = null
    private var instructionsByBackend: Map<String, String>? = null

    override val timeout: Long?
        get() = timeoutMs

    override fun getAllowedTools(): List<String> = listOf("AskUserQuestion")

    override fun getSystemPromptAppendix(): String {
        val provider = McpSystemPromptContext.getProvider()
        val key = when (provider) {
            AiAgentProvider.CODEX -> "codex"
            AiAgentProvider.CLAUDE -> "claude"
            else -> null
        }
        val override = key?.let { instructionsByBackend?.get(it) }?.trim().orEmpty()
        return if (override.isNotBlank()) override else DEFAULT_INSTRUCTIONS
    }

    fun setInstructionsByBackend(instructions: Map<String, String>?) {
        instructionsByBackend = instructions?.mapKeys { it.key.trim().lowercase() }
    }

    companion object {
        val DEFAULT_INSTRUCTIONS = """
When you need clarification from the user, especially when presenting multiple options or choices, use the MCP server `user-interaction` tool `AskUserQuestion` to ask questions.
Tool identifiers may differ across providers. Do not assume a fixed prefix or delimiter; select the tool that matches this server + tool pair.
The user's response will be returned to you through the same tool.
        """.trimIndent()

        val ASK_USER_QUESTION_SCHEMA: JsonObject = buildJsonObject {
            put("type", "object")
            put(
                "description",
                "Ask the user questions and get their choices. Use this tool to interact with users when input or confirmation is needed."
            )
            putJsonObject("properties") {
                putJsonObject("questions") {
                    put("type", "array")
                    put("description", "List of questions")
                    putJsonObject("items") {
                        put("type", "object")
                        putJsonObject("properties") {
                            putJsonObject("question") {
                                put("type", "string")
                                put("description", "Question content")
                            }
                            putJsonObject("header") {
                                put("type", "string")
                                put("description", "Question header/category label")
                            }
                            putJsonObject("options") {
                                put("type", "array")
                                put("description", "List of options")
                                putJsonObject("items") {
                                    put("type", "object")
                                    putJsonObject("properties") {
                                        putJsonObject("label") {
                                            put("type", "string")
                                            put("description", "Option display text")
                                        }
                                        putJsonObject("description") {
                                            put("type", "string")
                                            put("description", "Option description (optional)")
                                        }
                                    }
                                    putJsonArray("required") { add("label") }
                                }
                            }
                            putJsonObject("multiSelect") {
                                put("type", "boolean")
                                put("description", "Allow multiple selections, default false")
                            }
                        }
                        putJsonArray("required") {
                            add("question")
                            add("header")
                            add("options")
                        }
                    }
                }
            }
            putJsonArray("required") { add("questions") }
        }
    }

    fun setTimeoutMs(timeoutMs: Long?) {
        this.timeoutMs = timeoutMs
    }

    override suspend fun callToolJson(toolName: String, arguments: JsonObject): ToolResult {
        return when (toolName) {
            "AskUserQuestion" -> handleAskUserQuestionJson(arguments)
            else -> super.callToolJson(toolName, arguments)
        }
    }

    private suspend fun handleAskUserQuestionJson(arguments: JsonObject): ToolResult {
        val connectId = currentConnectId()
        val caller = connectId?.let { ClientCallerRegistry.get(it) }
            ?: return ToolResult.error("无法获取 ClientCaller，connectId=$connectId")

        mcpLogger.info { "📩 [AskUserQuestion] 收到工具调用，参数: $arguments" }

        val validationResult = SchemaValidator.validateWithSchema(
            schema = ASK_USER_QUESTION_SCHEMA,
            arguments = arguments,
            customValidators = listOf(
                { args ->
                    val questions = args["questions"]
                    if (questions is JsonPrimitive && questions.isString) {
                        com.asakii.server.mcp.schema.ValidationError(
                            parameter = "questions",
                            message = "Parameter should be an array, not a string",
                            hint = "Pass the array directly without serializing it to a string"
                        )
                    } else null
                }
            )
        )

        if (validationResult is ValidationResult.Invalid) {
            val errorMsg = validationResult.formatMessage()
            mcpLogger.warn { "⚠️ [AskUserQuestion] 参数校验失败:\n$errorMsg" }
            return ToolResult.error(errorMsg)
        }

        return try {
            val normalized = normalizeQuestions(arguments)
            val params = Json.decodeFromJsonElement(AskUserQuestionParams.serializer(), normalized)

            mcpLogger.info { "📤 [AskUserQuestion] 解析后的参数: ${params.questions.size} 个问题" }

            val protoRequest = AskUserQuestionRequest.newBuilder().apply {
                params.questions.forEach { q ->
                    addQuestions(ProtoQuestionItem.newBuilder().apply {
                        question = q.question
                        q.header?.let { header = it }
                        q.options?.forEach { opt ->
                            addOptions(ProtoQuestionOption.newBuilder().apply {
                                label = opt.label
                                if (opt.description.isNotEmpty()) {
                                    description = opt.description
                                }
                            }.build())
                        }
                        multiSelect = q.multiSelect
                    }.build())
                }
            }.build()

            val protoResponse = caller.callAskUserQuestion(protoRequest)

            mcpLogger.info { "📥 [AskUserQuestion] 收到前端响应: ${protoResponse.answersCount} 个回答" }

            // 构建回答映射
            val answersMap: Map<String, String> = protoResponse.answersList.associate {
                it.question to it.answer
            }

            // 生成 Markdown 格式的回复
            val content = buildString {
                appendLine("## User Answers")
                appendLine()
                params.questions.forEachIndexed { index, q ->
                    val answer = answersMap[q.question] ?: "(no answer)"
                    val header = q.header ?: "Question ${index + 1}"
                    appendLine("### $header")
                    appendLine("**Q:** ${q.question}")
                    appendLine("**A:** $answer")
                    appendLine()
                }
            }.trim()

            mcpLogger.info { "✅ [AskUserQuestion] 完成，返回:\n$content" }
            ToolResult.success(content)

        } catch (e: kotlinx.serialization.SerializationException) {
            mcpLogger.error { "❌ [AskUserQuestion] 参数格式错误: ${e.message}" }
            ToolResult.error("参数格式错误，请检查 questions 数组结构是否正确")
        } catch (e: Exception) {
            mcpLogger.error { "❌ [AskUserQuestion] 处理失败: ${e.message}" }
            e.printStackTrace()
            ToolResult.error("处理用户问题时发生错误，请稍后重试")
        }
    }

    private fun normalizeQuestions(arguments: JsonObject): JsonObject {
        val rawQuestions = arguments["questions"]
        if (rawQuestions is JsonPrimitive && rawQuestions.isString) {
            val content = rawQuestions.content
            if (content.startsWith("[") || content.startsWith("{")) {
                try {
                    val parsed = Json.parseToJsonElement(content)
                    if (parsed is JsonArray) {
                        return buildJsonObject {
                            arguments.forEach { (k, v) ->
                                if (k == "questions") put(k, parsed) else put(k, v)
                            }
                        }
                    }
                } catch (e: Exception) {
                    mcpLogger.warn { "⚠️ [AskUserQuestion] 无法从字符串解析 questions" }
                }
            }
        }
        return arguments
    }

    override suspend fun onInitialize() {
        registerToolWithSchema(
            name = "AskUserQuestion",
            description = "向用户询问问题并获取选择。使用此工具在需要用户输入或确认时与用户交互。",
            inputSchema = ASK_USER_QUESTION_SCHEMA
        ) { arguments ->
            handleAskUserQuestionJson(arguments)
        }

        mcpLogger.info { "✅ [UserInteractionMcpServer] 初始化完成，已注册 AskUserQuestion 工具" }
    }
}

File: vscode-extension/src/ide/mcp/userInteraction/userInteractionMcpServer.ts
/**
 * User Interaction MCP Server for VS Code
 * 
 * Provides AskUserQuestion tool for Claude to interact with users.
 * Uses RSocket to communicate with frontend Vue component.
 * 
 * 与 JetBrains 版本的 UserInteractionMcpServer.kt 完全对应。
 */

import { McpServer, RequestHandlerExtra } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { McpServerProvider, McpServerBase, createToolResult } from '../mcpServerRegistry';
import { ClientCallerRegistry } from '../../../server/rpc';
import { create } from '@bufbuild/protobuf';
import { Logger } from '../../../logging/logger';
import {
  AskUserQuestionRequestSchema,
  type AskUserQuestionRequest,
  type AskUserQuestionResponse,
} from '@proto/ide_pb';

// Custom AuthInfo type matching MCP SDK's AuthInfo with our extra field
interface McpAuthInfo {
    token: string;
    clientId: string;
    scopes: string[];
    extra?: {
        connectId?: string;
    };
}

// Logger instance
const logger = Logger.create('UserInteractionMcp');

// ========== Types ==========

/**
 * Question option
 */
interface QuestionOption {
    label: string;
    description?: string;
}

/**
 * Question item
 */
interface QuestionItem {
    question: string;
    header?: string;
    options?: QuestionOption[];
    multiSelect?: boolean;
}

/**
 * User answer item
 */
interface UserAnswerItem {
    question: string;
    header?: string;
    answer: string;
}

// ========== Zod Schemas ==========

const OptionSchema = z.object({
    label: z.string().describe('Option display text'),
    description: z.string().optional().describe('Option description (optional)')
});

const QuestionSchema = z.object({
    question: z.string().describe('Question content'),
    header: z.string().optional().describe('Question header/category label'),
    options: z.array(OptionSchema).optional().describe('List of options'),
    multiSelect: z.boolean().optional().default(false).describe('Allow multiple selections, default false')
});

const AskUserQuestionInputSchema = z.object({
    questions: z.array(QuestionSchema).describe('List of questions')
});

// ========== Default Instructions ==========

const DEFAULT_INSTRUCTIONS = `向用户询问问题并获取选择。使用此工具在需要用户输入或确认时与用户交互。`;

const SYSTEM_PROMPT_APPENDIX = `When you need clarification from the user, especially when presenting multiple options or choices, use the MCP server \`user_interaction\` tool \`AskUserQuestion\` to ask questions.
Tool identifiers may differ across providers. Do not assume a fixed prefix or delimiter; select the tool that matches this server + tool pair.
The user's response will be returned to you through the same tool.`;

// ========== User Interaction MCP Server ==========

/**
 * User Interaction MCP Server Implementation
 * 
 * Uses RSocket to communicate with frontend Vue component (AskUserQuestionInteractive.vue).
 * 与 JetBrains 版本架构完全一致。
 */
export class UserInteractionMcpServer extends McpServerBase {
    
    constructor() {
        super({
            name: 'user-interaction',
            version: '1.0.0',
            description: '用户交互工具服务器，提供向用户提问等功能'
        });
    }

    getSystemPromptAppendix(): string {
        return SYSTEM_PROMPT_APPENDIX;
    }

    getAllowedTools(): string[] {
        return ['AskUserQuestion'];
    }

    async initialize(): Promise<void> {
        // Register AskUserQuestion tool
        this.server.tool(
            'AskUserQuestion',
            DEFAULT_INSTRUCTIONS,
            // IMPORTANT:
            // McpServer.tool() expects a Zod "raw shape" as the params schema.
            // Passing a plain JSON schema object here will be treated as "annotations",
            // resulting in tool.inputSchema being EMPTY, which makes the model guess params.
            {
                questions: z.array(QuestionSchema).describe('List of questions')
            },
            async (params: { questions: QuestionItem[] }, extra: RequestHandlerExtra) => {
                // Get connectId from extra.authInfo.extra (passed from mcpHttpGateway via req.auth)
                logger.info(`Tool handler called! extra = ${JSON.stringify(extra, (key, value) => {
                    if (key === 'signal') return '[AbortSignal]';
                    return value;
                })}`);
                const authInfo = extra.authInfo as McpAuthInfo | undefined;
                logger.info(`authInfo = ${JSON.stringify(authInfo)}`);
                // connectId is now in authInfo.extra.connectId (MCP SDK AuthInfo structure)
                const connectId = authInfo?.extra?.connectId;
                logger.info(`connectId = ${connectId}`);
                return this.handleAskUserQuestion(params, connectId);
            }
        );

        logger.info('初始化完成，已注册 AskUserQuestion 工具');
    }

    /**
     * 处理 AskUserQuestion 工具调用
     * 
     * 通过 RSocket 调用前端 AskUserQuestionInteractive.vue 组件。
     * 
     * @param params 问题参数
     * @param connectId 连接ID，从 extra.authInfo 获取（由 mcpHttpGateway 通过 req.auth 传递）
     */
    private async handleAskUserQuestion(params: { questions: QuestionItem[] }, connectId?: string): Promise<{
        content: Array<{ type: 'text'; text: string }>;
        isError?: boolean;
    }> {
        logger.info(`收到 AskUserQuestion 调用, connectId=${connectId}`);
        
        // 列出所有已注册的 ClientCaller
        const allCallers = ClientCallerRegistry.getAll();
        logger.debug(`已注册的 ClientCaller: ${JSON.stringify(Array.from(allCallers.keys()))}`);
        
        const caller = connectId ? ClientCallerRegistry.get(connectId) : undefined;

        if (!caller) {
            const errorMsg = `无法获取 ClientCaller，connectId=${connectId}, 已注册: [${Array.from(allCallers.keys()).join(', ')}]`;
            logger.warn(errorMsg);
            return createToolResult(errorMsg, true);
        }

        logger.info(`参数: ${JSON.stringify(params)}`);

        try {
            // 验证参数
            const parsed = AskUserQuestionInputSchema.safeParse(params);
            if (!parsed.success) {
                const errorMsg = `参数校验失败: ${parsed.error.message}`;
                logger.warn(errorMsg);
                return createToolResult(errorMsg, true);
            }

            const { questions } = parsed.data;
            logger.info(`解析后的参数: ${questions.length} 个问题`);

            // 构建 Protobuf 请求
            const protoRequest = create(AskUserQuestionRequestSchema, {
                questions: questions.map(q => ({
                    question: q.question,
                    header: q.header || '',
                    multiSelect: q.multiSelect || false,
                    options: (q.options || []).map(opt => ({
                        label: opt.label,
                        description: opt.description || ''
                    }))
                }))
            });

            logger.info('正在调用前端 callAskUserQuestion...');
            
            // 通过 RSocket 调用前端
            const protoResponse = await caller.callAskUserQuestion(protoRequest);

            logger.info(`收到前端响应: ${protoResponse.answers.length} 个回答`);

            // 构建回答映射
            const answersMap: Map<string, string> = new Map();
            for (const answer of protoResponse.answers) {
                answersMap.set(answer.question, answer.answer);
            }

            // 生成 Markdown 格式的回复
            let content = '## User Answers\n\n';
            questions.forEach((q, index) => {
                const answer = answersMap.get(q.question) || '(no answer)';
                const header = q.header || `Question ${index + 1}`;
                content += `### ${header}\n`;
                content += `**Q:** ${q.question}\n`;
                content += `**A:** ${answer}\n\n`;
            });

            logger.info(`完成，返回:\n${content.trim()}`);
            return createToolResult(content.trim());

        } catch (e) {
            const errorMsg = e instanceof Error ? e.message : String(e);
            logger.error(`处理失败: ${errorMsg}`);
            return createToolResult(`处理用户问题时发生错误: ${errorMsg}`, true);
        }
    }
}

// ========== Provider ==========

/**
 * User Interaction MCP Server Provider
 */
export class UserInteractionMcpServerProvider implements McpServerProvider {
    name = 'user-interaction';
    private server: UserInteractionMcpServer;

    constructor() {
        this.server = new UserInteractionMcpServer();
    }

    async initialize(): Promise<void> {
        await this.server.initialize();
    }

    getServer(): McpServer {
        return this.server.getServer();
    }

    getDisallowedBuiltinTools(): string[] {
        return [];
    }

    dispose(): void {
        // Cleanup if needed
    }
}
